modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/JavaHelidonCommonCodegen.java

```diff
@@ -321,4 +368,157 @@ private String chooseExampleReturnTypeValue(CodegenOperation op) {
+            }
+            return INSTANCE;
+        }
+
+        private final List<String> versions;
+
+        private VersionUtil() throws BackingStoreException, IOException {
+            versions = versions();
+        }
+
+        String defaultVersion() {
+            return versions.get(0);
+        }
+
+        static int majorVersion(String fullVersion) {
+            return Integer.parseUnsignedInt(fullVersion.substring(0, fullVersion.indexOf('.')));
+        }
+
+        /**
+         * Returns the version that is the "closest" match to the requested version prefix.
+         *
+         * @param requestedVersionPrefix prefix to use in matching versions
+         * @return matching version
+         */
+        String chooseVersion(String requestedVersionPrefix) {
+            String result = null;
+            while (result == null && !requestedVersionPrefix.isBlank()) {
+                result = matchVersion(requestedVersionPrefix, versions);
+                // Trim requested version from the back by "dots" to try to find an expression that matches.
+                int lastDot = requestedVersionPrefix.lastIndexOf('.');
+
+                requestedVersionPrefix = lastDot == -1 ? "" : requestedVersionPrefix.substring(0, lastDot);
+            }
+            return result;
+        }
+
+        /**
+         * Returns the last version that matches the provided prefix.
+         *
+         * @param versionPrefix prefix to use in matching against actual versions
+         * @param descendingOrderVersions supported versions in descending order
+         * @return specific version matching the prefix; null if none
+         */
+        private static String matchVersion(String versionPrefix, List<String> descendingOrderVersions) {
+            for (String candidate : descendingOrderVersions) {
+                if (candidate.startsWith(versionPrefix)) {
+                    return candidate;
+                }
+            }
+            return null;
+        }
+
+        /**
+         * Retrieves the list of supported versions from the web site or, failing that, a local file.
+         *
+         * @return list of supported versions
+         * @throws IOException in case of error accessing the web site and reading the local file
+         */
+        private static List<String> versions() throws IOException, BackingStoreException {
+
+            HttpClient httpClient = HttpClient.newBuilder()
+                    .connectTimeout(CONNECTION_TIMEOUT)
+                    .build();
+            try {
+
+                HttpRequest req = HttpRequest.newBuilder(URI.create(HELIDON_VERSIONS_URL))
+                        .GET()
+                        .timeout(REQUEST_TIMEOUT)
+                        .build();
+
+                HttpResponse<String> response = httpClient.send(req, HttpResponse.BodyHandlers.ofString());
+
+                Preferences versionPrefs = Preferences.userRoot().node(PREFERENCES_KEY);
```

Comment:
The Helidon CLI caches `versions.xml` in `~/.helidon/cache/versions.xml`. We should at least discuss if using that mechanism here makes sense -- I see pros and cons.

Reply:
Yeah, I knew that the CLI caches the data. 

My thinking about using `Preferences` was mostly that as a Java feature, not a Helidon-specific one, using it might cause less friction in getting the PR approved. Eventually the generators will likely use the yet-to-be-designed Helidon-to-MP-release mapping file that would be different from `versions.xml` in case that makes any difference.